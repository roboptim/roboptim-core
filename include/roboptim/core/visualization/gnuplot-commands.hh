// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_VISUALIZATION_GNUPLOT_COMMANDS_HH
# define ROBOPTIM_CORE_VISUALIZATION_GNUPLOT_COMMANDS_HH
# include <roboptim/core/sys.hh>
# include <roboptim/core/debug.hh>

# include <string>
# include <iostream>

namespace roboptim
{
  namespace visualization
  {
    namespace gnuplot
    {
      /// \addtogroup roboptim_visualization
      /// @{

      /// \brief Gnuplot command.
      ///
      /// A command represents a Gnuplot command.
      /// The representation of the command as a string
      /// can be accessed through the #command() method.
      class ROBOPTIM_DLLAPI Command
      {
      public:
	/// \brief Make a command from a string.
	explicit Command (std::string cmd) throw ();

	~Command () throw ();

	/// \brief Retrieve the command as a string.
	const std::string& command () const throw ();
      protected:
	/// \brief Store Gnuplot command.
	std::string command_;
      };

      /// \brief Make a Gnuplot comment.
      ROBOPTIM_DLLAPI Command comment (const char*) throw ();

      /// \brief Make a Gnuplot comment.
      template <typename T>
      ROBOPTIM_DLLAPI Command comment (const T& content) throw ()
      {
          // Note: we do not use boost::lexical_cast because the << operators
          // need to be in the std:: or boost:: namespaces. As a result, if we
          // try to add a comment with an Eigen matrix, it will not be printed
          // in the RobOptim way. Thus, we stick to stringstream (for now).
          std::stringstream ss;
          ss << "# " << content;
          return Command (ss.str ());
      }

      /// \brief Make a Gnuplot set command.
      ///
      /// Make a Gnuplot set command (assign a new
      /// value to a variable in Gnuplot).
      /// \param var variable that will be set
      /// \param val new variable value
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command set (const char* var, const char* val = "") throw ();

      /// \brief Make a Gnuplot unset command.
      ///
      /// Make a Gnuplot unset command (destroy a variable).
      /// \param var variable that will be unset
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command unset (const char* var) throw ();

      /// \brief Make a Gnuplot show command.
      ///
      /// Make a Gnuplot show command (show a variable value).
      /// \param var variable that will be shown
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command show (const char* var) throw ();

      /// \brief Make a Gnuplot clear command.
      ///
      /// Make a Gnuplot clear command (clear the screen).
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command clear () throw ();

      /// \brief Make a Gnuplot pwdcommand.
      ///
      /// Make a Gnuplot pwd command (show the current directory).
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command pwd () throw ();

      /// \brief Make a Gnuplot quit command.
      ///
      /// Make a Gnuplot quit command (quit Gnuplot).
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command quit () throw ();

      /// \brief Make a Gnuplot replot command.
      ///
      /// Make a Gnuplot replot command (redraw previously defined curves).
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command replot () throw ();

      /// \brief Make a Gnuplot reread command.
      ///
      /// Make a Gnuplot reread command (reread data files).
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command reread () throw ();

      /// \brief Make a Gnuplot reset command.
      ///
      /// Make a Gnuplot replot command (reset Gnuplot state).
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command reset () throw ();

      /// \brief Make a Gnuplot cd command.
      ///
      /// Make a Gnuplot cd command (change current directory).
      /// \param dir new directory
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command cd (const char* dir) throw ();

      /// \brief Make a Gnuplot help command.
      ///
      /// Make a Gnuplot help command (display Gnuplot help).
      /// \param topic help topic
      /// \return Gnuplot command
      ROBOPTIM_DLLAPI Command help (const char* topic = "") throw ();

      /// @}

    } // end of namespace gnuplot.
  } // end of namespace visualization.
} // end of namespace roboptim.

#endif //! ROBOPTIM_CORE_VISUALIZATION_GNUPLOT_COMMANDS_HH
