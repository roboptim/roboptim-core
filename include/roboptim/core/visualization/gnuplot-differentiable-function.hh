// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_VISUALIZATION_GNUPLOT_DIFFERENTIABLE_FUNCTION_HH
# define ROBOPTIM_CORE_VISUALIZATION_GNUPLOT_DIFFERENTIABLE_FUNCTION_HH
# include <boost/mpl/assert.hpp>
# include <roboptim/core/sys.hh>
# include <roboptim/core/debug.hh>

# include <roboptim/core/function.hh>
# include <roboptim/core/differentiable-function.hh>

namespace roboptim
{
  namespace visualization
  {
    namespace gnuplot
    {
      /// \addtogroup roboptim_visualization
      /// @{

      /// \brief Plot the Jacobian structure with Gnuplot.
      ///
      /// Plot the structure of the Jacobian of a differentiable function with
      /// Gnuplot. Nonzero values will be displayed in blue, zeros in white.
      ///
      /// When dealing with a dense matrix, the actual values are printed and
      /// exact zeros will be displayed in white.
      ///
      /// When dealing with a sparse matrix, inserted zeros will also be treated
      /// as nonzero values, thus returning the real structure of the sparse
      /// Jacobian matrix. In this case, the values returned are 1 for sparse
      /// elements, and 0 for actual zeros.
      ///
      /// \param f differentiable function whose Jacobian will be plotted
      /// \param arg optimization parameters of the point to plot. This
      /// parameter is not relevant when dealing with sparse functions, since
      /// it should not change the Jacobian's sparse structure.
      /// \return Gnuplot command
      template <typename T>
      Command plot_jac (const GenericDifferentiableFunction<T>& f,
                        const typename GenericDifferentiableFunction<T>::
                        argument_t& arg);

      template <>
      ROBOPTIM_DLLAPI
      Command plot_jac (const DifferentiableFunction& f,
                        const DifferentiableFunction::argument_t& arg);
      template <>
      ROBOPTIM_DLLAPI
      Command plot_jac (const DifferentiableSparseFunction& f,
                        const DifferentiableSparseFunction::argument_t& arg);

      template <typename T>
      Command plot_jac (const GenericDifferentiableFunction<T>&,
                        const typename GenericDifferentiableFunction<T>::
                        argument_t&)
      {
	BOOST_MPL_ASSERT_MSG (false, NOT_IMPLEMENTED, ());
	return Command ("");
      }

      /// @}
    } // end of namespace gnuplot.
  } // end of namespace visualization.
} // end of namespace roboptim.

#endif //! ROBOPTIM_CORE_VISUALIZATION_GNUPLOT_DIFFERENTIABLE_FUNCTION_HH
