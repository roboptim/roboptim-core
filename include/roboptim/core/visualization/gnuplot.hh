// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_VISUALIZATION_GNUPLOT_HH
# define ROBOPTIM_CORE_VISUALIZATION_GNUPLOT_HH
# include <roboptim/core/sys.hh>
# include <roboptim/core/debug.hh>
# include <roboptim/core/function.hh>

# include <vector>

# define EIGEN_YES_I_KNOW_SPARE_MODULE_IS_NOT_STABLE_YET
# include <Eigen/Core>
# include <Eigen/Sparse>

# include <roboptim/core/function.hh>
# include <roboptim/core/visualization/fwd.hh>
# include <roboptim/core/visualization/gnuplot-commands.hh>

namespace roboptim
{
  namespace visualization
  {
    /// \addtogroup roboptim_visualization
    /// @{

    /// \brief Normalize floating point number output.
    /// Normalization is done on all displayed floating
    /// point numbers to get a consistent output.
    double normalize (const double& x);

    /// \brief Apply normalize to each element of a matrix.
    Eigen::MatrixXd
    normalize (const Eigen::MatrixXd& x);


    /// \brief Apply normalize to each element of a container.
    template <typename T>
    T normalize (const T& x);

    inline double
    normalize (const double& x)
    {
      if (std::fabs (x) < 1e-8)
	return 0.;
      return x;
    }

    inline Eigen::MatrixXd
    normalize (const Eigen::MatrixXd& x)
    {
      Eigen::MatrixXd res (x.rows (), x.cols ());
      for (Function::size_type i = 0; i < x.rows (); ++i)
	for (Function::size_type j = 0; j < x.cols (); ++j)
	  res (i, j) = normalize (x (i, j));
      return res;
    }

    template <typename T>
    T
    normalize (const T& x)
    {
      T res (x.size ());
      for (Function::size_type i = 0; i < x.size (); ++i)
	res[i] = normalize (x[i]);
      return res;
    }


    /// \brief Gnuplot script
    ///
    /// This class gathers Gnuplot commands
    /// to build a complete Gnuplot script.
    /// Gnuplot commands can be inserted through
    /// this object through the \c << operator
    /// and this object can be put into an output stream
    /// using the \c << operator.
    class ROBOPTIM_DLLAPI Gnuplot
    {
    public:
      ~Gnuplot () throw ();

      /// \brief Instanciate a Gnuplot without setting a term.
      /// \param with_header whether to print the header or not
      /// \return Gnuplot instance
      static Gnuplot make_gnuplot (bool with_header = true) throw ()
      {
	return Gnuplot (with_header);
      }

      /// \brief Instanciate a Gnuplot suitable for interactive use.
      ///
      /// This initializes a persistent Gnuplot instance which are
      /// suitable for user interaction.
      /// \param with_header whether to print the header or not
      /// \return Gnuplot instance
      static Gnuplot make_interactive_gnuplot (bool with_header = true) throw ()
      {
	Gnuplot gp(with_header);
	gp.push_command (gnuplot::set ("terminal", "wxt persist"));
	return gp;
      }

      /// \brief Add a new Gnuplot command to the script.
      /// \param cmd command that will be pushed
      void push_command (gnuplot::Command cmd) throw ();


      /// \brief Display the Gnuplot script on the specified output stream.
      ///
      /// \param o output stream used for display
      /// \return output stream
      std::ostream& print (std::ostream&) const throw ();

      Gnuplot& operator << (gnuplot::Command) throw ();

    protected:
      /// \brief Default constructor can not be called directly.
      ///
      /// Use of the named constructor (see static methods) to
      /// instantiate this class.
      explicit Gnuplot (bool with_header = true) throw ();
    private:
      /// \brief Vector of commands.
      std::vector<gnuplot::Command> commands_;
    };

    /// Example shows simple Gnuplot visualization.
    /// \example visualization-gnuplot-simple.cc

    /// Example shows function display with Gnuplot.
    /// \example visualization-gnuplot-function.cc

    /// \brief Override operator<< to handle Gnuplot command insertion.
    ///
    /// \tparam Gnuplot command type
    /// \param gp Gnuplot script that will receive the new command
    /// \param t new Gnuplot command.
    /// \return modified Gnuplot script
    template <typename T>
    Gnuplot& operator<< (Gnuplot& gp, T t)
    {
      gp.push_command (t ());
      return gp;
    }

    /// @}

    /// \brief Override operator<< to handle Gnuplot script display.
    ///
    /// \param o output stream used for display
    /// \param gp Gnuplot script to be displayed
    /// \return output stream
    ROBOPTIM_DLLAPI std::ostream& operator<< (std::ostream&, const Gnuplot& gp);
  } // end of namespace visualization.
} // end of namespace roboptim.

#endif //! ROBOPTIM_CORE_VISUALIZATION_GNUPLOT_HH
