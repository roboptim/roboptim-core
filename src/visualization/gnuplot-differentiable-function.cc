// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include <roboptim/core/visualization/gnuplot.hh>
#include <roboptim/core/visualization/gnuplot-commands.hh>
#include <roboptim/core/visualization/gnuplot-matrix.hh>
#include <roboptim/core/visualization/gnuplot-differentiable-function.hh>

# include <boost/format.hpp>
# include <boost/lexical_cast.hpp>

namespace roboptim
{
  namespace visualization
  {
    namespace gnuplot
    {
      namespace detail
      {

        std::string dense_jacobian_to_gnuplot
        (const DifferentiableFunction::jacobian_t& jac,
         const std::string& name)
        {
          std::string str = "set title 'Dense Jacobian (" + name + ")'\n";
          str += plot_mat (jac).command ();

          return str;
        }


        std::string sparse_jacobian_to_gnuplot
        (const DifferentiableSparseFunction::jacobian_t& jac,
         const std::string& name)
        {
          std::string str = "set title 'Sparse Jacobian (" + name + ")'\n";
          str += plot_mat (jac).command ();

          return str;
        }

      } // end of namespace detail.

      template <>
      Command plot_jac (const DifferentiableFunction& f,
                        const DifferentiableFunction::argument_t& arg)
      {
	DifferentiableFunction::jacobian_t jac = f.jacobian(arg);

        return Command (detail::dense_jacobian_to_gnuplot
                        (jac, f.getName()));
      }


      template <>
      Command plot_jac (const DifferentiableSparseFunction& f,
                        const DifferentiableSparseFunction::argument_t& arg)
      {
        DifferentiableSparseFunction::jacobian_t jac = f.jacobian(arg);

        return Command (detail::sparse_jacobian_to_gnuplot
                        (jac, f.getName()));
      }

    } // end of namespace gnuplot.
  } // end of namespace visualization.
} // end of namespace roboptim
