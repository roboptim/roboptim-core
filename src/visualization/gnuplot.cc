// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "roboptim/core/config.hh"
#include "config.h"

#include <iostream>

#include <roboptim/core/visualization/gnuplot.hh>
#include <roboptim/core/visualization/gnuplot-commands.hh>
#include <roboptim/core/indent.hh>

namespace roboptim
{
  namespace visualization
  {
    Gnuplot::Gnuplot (bool with_header) throw ()
    {
      using namespace gnuplot;
      push_command (comment ("!/usr/bin/gnuplot"));
      if(with_header) push_command (comment (" Generated by " PACKAGE_NAME));
    }

    Gnuplot::~Gnuplot () throw ()
    {
    }

    void
    Gnuplot::push_command (gnuplot::Command cmd) throw ()
    {
      commands_.push_back (cmd);
    }

    std::ostream&
    Gnuplot::print (std::ostream& o) const throw ()
    {
      typedef std::vector<gnuplot::Command>::const_iterator citer_t;

      for (citer_t it = commands_.begin (); it != commands_.end (); ++it)
	o << it->command () << std::endl;
      return o;
    }

    Gnuplot&
    Gnuplot::operator << (gnuplot::Command cmd) throw ()
    {
      push_command (cmd);
      return *this;
    }

    std::ostream&
    operator<< (std::ostream& o, const Gnuplot& gp)
    {
      return gp.print (o);
    }
  } // end of namespace visualization.
} // end of namespace roboptim.
