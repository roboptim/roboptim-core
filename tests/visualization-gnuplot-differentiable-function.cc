// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "shared-tests/common.hh"

#include <iostream>

#include <roboptim/core/io.hh>

#include <roboptim/core/visualization/gnuplot.hh>
#include <roboptim/core/visualization/gnuplot-commands.hh>
#include <roboptim/core/visualization/gnuplot-differentiable-function.hh>

using namespace roboptim;
using namespace roboptim::visualization;


// Define f(x) = forty_two(x)
struct Square : public DifferentiableFunction
{
  explicit Square ()
    : DifferentiableFunction (7, 7, "The Answer")
  {
  }

  void impl_compute (result_t& result,
		     const argument_t& argument) const throw ()
  {
    result[0] = argument[0] + argument[4] + argument[5];
    result[1] = argument[0] + argument[2] + argument[6];
    result[2] = argument[0] + argument[2] + argument[6];
    result[3] = argument[0] + argument[1] + argument[2] + argument[5];
    result[4] = argument[2] + argument[4];
    result[5] = argument[2] + argument[4];
    result[6] = argument[2] + argument[4] + argument[5] + argument[6];
  }

  void impl_gradient (gradient_t&, const argument_t&,  size_type)
    const throw ()
  {
  }

  void impl_jacobian (jacobian_t& jac, const argument_t&)
    const throw ()
  {
    jac.setZero();
    jac(0,0) = 1.0;
    jac(0,4) = 1.0;
    jac(0,5) = 1.0;
    jac(1,0) = 1.0;
    jac(1,2) = 1.0;
    jac(1,6) = 1.0;
    jac(2,0) = 1.0;
    jac(2,2) = 1.0;
    jac(2,6) = 1.0;
    jac(3,0) = 1.0;
    jac(3,1) = 1.0;
    jac(3,2) = 1.0;
    jac(3,5) = 1.0;
    jac(4,2) = 1.0;
    jac(4,4) = 1.0;
    jac(5,2) = 1.0;
    jac(5,4) = 1.0;
    jac(6,2) = 1.0;
    jac(6,4) = 1.0;
    jac(6,5) = 1.0;
    jac(6,6) = 1.0;
  }
};


BOOST_AUTO_TEST_CASE (visualization_gnuplot_differentiable_function)
{
  boost::shared_ptr<boost::test_tools::output_test_stream>
    output = retrievePattern ("visualization-gnuplot-differentiable-function");

  using namespace roboptim::visualization::gnuplot;
  Gnuplot gnuplot = Gnuplot::make_interactive_gnuplot ();

  Square square;
  Square::vector_t arg(7);
  arg.fill(1.0);

  (*output)
    << (gnuplot
	<< plot_jac (square, arg)
	);

  std::cout << output->str () << std::endl;
  BOOST_CHECK (output->match_pattern ());
}
