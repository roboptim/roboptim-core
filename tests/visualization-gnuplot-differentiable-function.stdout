#!/usr/bin/gnuplot
# Generated by roboptim-core
set terminal wxt persist
set multiplot layout 2,1
# Dense matrix
set title 'Dense Jacobian (The Answer)'
set palette defined(0 "white",1 "blue")
set grid front
set xrange [0:7]
set yrange [0:7] reverse
set size ratio -1
unset colorbox
plot '-' using ($1+0.5):($2+0.5):($3 == 0 ? 0 : 1) matrix with image notitle
1.00000000 0.00000000 0.00000000 0.00000000 1.00000000 1.00000000 0.00000000
1.00000000 0.00000000 1.00000000 0.00000000 0.00000000 0.00000000 1.00000000
1.00000000 0.00000000 1.00000000 0.00000000 0.00000000 0.00000000 1.00000000
1.00000000 1.00000000 1.00000000 0.00000000 0.00000000 1.00000000 0.00000000
0.00000000 0.00000000 1.00000000 0.00000000 1.00000000 0.00000000 0.00000000
0.00000000 0.00000000 1.00000000 0.00000000 1.00000000 0.00000000 0.00000000
0.00000000 0.00000000 1.00000000 0.00000000 1.00000000 1.00000000 1.00000000
e
e

# Sparse matrix
set title 'Sparse Jacobian (The Answer)'
set palette defined(0 "white",1 "blue")
set grid front
set xrange [0:7]
set yrange [0:7] reverse
set size ratio -1
unset colorbox
plot '-' using ($1+0.5):($2+0.5):($3 == 0 ? 0 : 1) matrix with image notitle
1 0 0 0 1 1 0
1 0 1 0 0 0 1
1 0 1 0 0 0 1
1 1 1 0 0 1 0
0 0 1 0 1 0 0
0 0 1 0 1 0 0
0 0 1 0 1 1 1
e
e

unset multiplot
